app/cdap/components/hydrator/components/Canvas/index.tsx

```diff
@@ -0,0 +1,563 @@
+
+// This is to overwrite the styles of pipeline comments button
+const StyledControlButton = styled(ControlButton)`
+  button {
+    padding: 0;
+    border: none;
+    width: 25px;
+    height: 25px;
+  }
+`;
+
+const getConnectionsForDisplay = (connections, nodes) => {
+  return connections.map((conn) => {
+    const reactFlowConn: any = {
+      id: 'reactflow__edge-' + conn.from + '-' + conn.to,
+      source: conn.from,
+      target: conn.to,
+      ...EdgeStyle,
+    };
+    const toNode = nodes.find((node) => node.name === conn.to);
+    if (toNode.type === PLUGIN_TYPES.ERROR_TRANSFORM) {
+      reactFlowConn.sourceHandle = 'source_error';
+    } else if (toNode.type === PLUGIN_TYPES.ALERT_PUBLISHER) {
+      reactFlowConn.sourceHandle = 'source_alert';
+    }
+    return reactFlowConn;
+  });
+};
+
+const Canvas = ({
+  angularNodes,
+  angularConnections,
+  isDisabled,
+  previewMode,
+  updateNodes,
+  updateConnections,
+  onPropertiesClick,
+  onMetricsClick,
+  getAngularConnections,
+  getAngularNodes,
+  setSelectedNodes,
+  setSelectedConnections,
+  onKeyboardCopy,
+  setPluginActiveForComment,
+  getActivePluginForComment,
+  setPluginComments,
+  getPluginConfiguration,
+  getCustomIconSrc,
+  shouldShowAlertsPort,
+  shouldShowErrorsPort,
+  undoActions,
+  redoActions,
+  pipelineComments,
+  setPipelineComments,
+  onPreviewData,
+  cleanUpGraph,
+  pipelineArtifactType,
+  metricsData,
+  metricsDisabled,
+}: ICanvasProps) => {
+  const reactFlowInstance = useReactFlow();
+  const deletePressed = useKeyPress(['Backspace', 'Delete']);
+  const copyPressed = useKeyPress(['Meta+c', 'Ctrl+c']);
+  const pastePressed = useKeyPress(['Meta+v', 'Ctrl+v']);
+  const [copiedElements, setCopiedElements] = useState<{
+    nodes: any[];
+    edges: any[];
+  }>({
+    nodes: [],
+    edges: [],
+  });
+  const [selectedElements, setSelectedElements] = useState<{
+    nodes: any[];
+    edges: any[];
+  }>({
+    nodes: [],
+    edges: [],
+  });
+
+  const getNodesForDisplay = (nodes, reactNodes, isCleanUp = false) => {
+    let existingIds = [];
+    if (!isCleanUp) {
+      existingIds = reactNodes.map((node) => node.id);
+    }
+    return nodes.map((node) => {
+      const data = {
+        node,
+        onPropertiesClick,
+        onMetricsClick,
+        setPluginActiveForComment,
+        getActivePluginForComment,
+        setPluginComments,
+        getSelectedConnections,
+        getSelectedNodes,
+        getPluginConfiguration,
+        getCustomIconSrc,
+        shouldShowAlertsPort,
+        shouldShowErrorsPort,
+        previewMode,
+        onPreviewData,
+        copySelectedNodeId,
+        deleteSelectedNodeId,
+        isDisabled,
+        metricsData,
+        metricsDisabled,
+      };
+      const reactflowNode = {
+        id: node.name,
+        data,
+        type: 'plugin',
+        position: {
+          x: parseInt(node._uiPosition.left, 10),
```

Comment:
See https://google.github.io/styleguide/tsguide.html#type-coercion - use `Number(node._Position.foo)` and check for `NaN`

Reply:
it's not pure number string ```node._uiPosition
{top: '150px', left: '107.9px'}``` 
What's the best way to handle?